compiler/ast.nim

```diff
@@ -331,6 +331,8 @@ type
     nfHasComment # node has a comment
     nfSkipFieldChecking # node skips field visable checking
     nfOpenSym # node is a captured sym but can be overriden by local symbols
+              # ideally a unary node containing nkSym/nkOpenSymChoice or an
+              # extension over nkOpenSymChoice
```

Comment:
To elaborate on this (extension over `nkOpenSymChoice`):

If some version of #23104 is implemented (IMO the node flag for this instead is the lesser of 2 evils, but changing the representation for it is easier anyway), we could replace every `nkSym` instance of `nfOpenSym` with a unary `nkOpenSymChoice` that always prefers its first node, then add the symbol-replacing logic to `nkOpenSymChoice` in general, so we don't need `nfOpenSym` anymore. The condition for the new symbol replacing the symchoice would then become `s2 != symChoice.preferredSym and s2 notin OverloadableSyms`. The only problem is making sure the compiler can handle `nkOpenSymChoice` instead of `nkSym` in those spots.

The reason I couldn't implement these in a merged manner before is because I couldn't think of the symbol replacing logic we have now that only checks symbols in the generic proc instantiation.

Reply:
Give a new node kind a try please, I consider every single flag a hack that lives on for far too long.

Reply:
I'll try it in another PR, I think the merging with `nkOpenSymChoice` should still eventually be done so we might end up deprecating the new node kind, at least maybe before the experimental switch is default.

I'm also not sure about the best representation for #23104, this seems wrong to me:

```nim
OpenSymChoice
  Preferred
    Sym "foo"
  Sym "foo"
  ...
```

Adding `nkPreferredOpenSymChoice` and `nkPreferredClosedSymChoice` seems kind of excessive too.

An option that isn't as intuitive but should be easier for the compiler/macros to deal with is to leave the first child of the symchoice empty/as an ident node if there isn't a preferred choice, and to make the preferred choice the first child if there is one.

```nim
# no preference:
OpenSymChoice
  Empty # or Ident "foo"
  Sym "foo"
  Sym "foo"

# first Sym "foo" is preferred:
OpenSymChoice
  Sym "foo"
  Sym "foo"
  ...
```